### PR TITLE
`com.github.rhyme_player.rhyme`

### PR DESCRIPTION
**Pull request by Abdallah-Moh**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

File: com.github.rhyme_player.rhyme.yml
Comment (hfiguiere) [inline com.github.rhyme_player.rhyme.yml:66]:
```suggestion
        path: com.github.rhyme_player.rhyme.desktop
```

File: com.github.rhyme_player.rhyme.yml
Comment (hfiguiere) [inline com.github.rhyme_player.rhyme.yml:68]:
```suggestion
        path: com.github.rhyme_player.rhyme.appdata.xml
```

**Comment by hfiguiere** [comment]
bot, build com.github.rhyme_player.rhyme

File: com.github.rhyme_player.rhyme.appdata.xml
Comment (hfiguiere) [inline com.github.rhyme_player.rhyme.appdata.xml:3]: this wont validate: missing releases, project license, oars, screenshots. at least.

**Review by hfiguiere — CHANGES_REQUESTED** [review]

**Comment by hfiguiere** [comment]
bot, build com.github.rhyme_player.rhyme

**Comment by hfiguiere** [comment]
Other reviewers might have comment, and then admin will merge if they think it's good to, or ask questions / request changes. After they merge, it will be available and you'll have access to the repository to maintain it.

File: com.github.rhyme_player.rhyme.appdata.xml
Comment (hfiguiere) [inline com.github.rhyme_player.rhyme.appdata.xml:5]: the validator didn't complain, but we also want to have the `project_license` tag here, to indicate which license the application is under.

File: com.github.rhyme_player.rhyme.yml
Comment (hfiguiere) [inline com.github.rhyme_player.rhyme.yml:6]: one more thing: there is discrepency between the base app and the runtime.

GIven that https://github.com/flathub/org.electronjs.Electron2.BaseApp/issues/29 is open, I think the runtime version should be 20.08 for now.

**Comment by hfiguiere** [comment]
bot, build com.github.rhyme_player.rhyme

File: com.github.rhyme_player.rhyme.yml
Comment (gasinvein) [inline com.github.rhyme_player.rhyme.yml:69]: Nitpick: apps on Flathub usually pup build-commands above `sources` and below `build-options`.

File: com.github.rhyme_player.rhyme.yml
Comment (gasinvein) [inline com.github.rhyme_player.rhyme.yml:75]: This `jq` command doesn't depend on on env vars from `electron-builder-arch-args.sh`, thus no need to concatenate it with `yarn build` command. Besides, shell will eat its exit code this way. Better put it in its own `build-commands` array item.
```suggestion
      - jq '.build.linux.target="dir"' <<<$(<package.json) > package.json
      - |
        . ../flatpak-node/electron-builder-arch-args.sh
        yarn build && yarn electron-builder -- $ELECTRON_BUILDER_ARCH_ARGS  --linux --dir
```

File: com.github.rhyme_player.rhyme.appdata.xml
Comment (gasinvein) [inline com.github.rhyme_player.rhyme.appdata.xml:1]: Metainfo isn't specific to flatpak, deb and rpm repositories can benefit from it as well. Consider moving it to your upstream repository.

File: com.github.rhyme_player.rhyme.yml
Comment (gasinvein) [inline com.github.rhyme_player.rhyme.yml:24]: Yarn was added to the nodejs SDK extension, you don't need to build it yourself.

Comment (Abdallah-Moh) [reply com.github.rhyme_player.rhyme.yml:24]: do I need to set in environment variables for that

Comment (gasinvein) [reply com.github.rhyme_player.rhyme.yml:24]: No, just remove the `yarn` module.

File: com.github.rhyme_player.rhyme.appdata.xml
Comment (Abdallah-Moh) [reply com.github.rhyme_player.rhyme.appdata.xml:1]: How can I tell electron-builder use them

**Comment by Abdallah-Moh** [comment]
bot, build com.github.rhyme_player.rhyme

File: com.github.rhyme_player.rhyme.appdata.xml
Comment (gasinvein) [reply com.github.rhyme_player.rhyme.appdata.xml:1]: No idea, sorry. Probably it can run some custom commands at deb/rpm package build stage, or has other way to add a custom file to the resulting package?

Comment (Abdallah-Moh) [reply com.github.rhyme_player.rhyme.appdata.xml:1]: I will check it out thanks anyway

File: com.github.rhyme_player.rhyme.yml
Comment (gasinvein) [inline com.github.rhyme_player.rhyme.yml:31]: Since you're using Yarn, these environment variables shouldn't be needed.

File: com.github.rhyme_player.rhyme.yml
Comment (gasinvein) [inline com.github.rhyme_player.rhyme.yml:16]: Consider limiting filesystem permissions to `xdg-music` (and probably adding removable media directories, i.e. `/media` and `/run/media`).

Comment (Abdallah-Moh) [reply com.github.rhyme_player.rhyme.yml:16]: The app can change the music directory to any where else on the os

Comment (Abdallah-Moh) [reply com.github.rhyme_player.rhyme.yml:16]: This will limit it only to the music folder

Comment (gasinvein) [reply com.github.rhyme_player.rhyme.yml:16]: > This will limit it only to the music folder

Yes, that's the point. Apps should try to stay secure-by-default, while full `home` access allows trivial sandbox escape. If your app doesn't modify the music directory, you can at least limit home access to read-only (changing this to `home:ro`).

Comment (Abdallah-Moh) [reply com.github.rhyme_player.rhyme.yml:16]: ok

Comment (Abdallah-Moh) [reply com.github.rhyme_player.rhyme.yml:16]: I will change it to read-only

**Comment by nedrichards** [comment]
bot, build com.github.rhyme_player.rhyme

File: com.github.rhyme_player.rhyme.yml
Comment (nedrichards) [inline com.github.rhyme_player.rhyme.yml:3]: 21.08 is just out - could you try running with that instead? It'll obviously be in security support longer and should otherwise be newer and better.

File: com.github.rhyme_player.rhyme.yml
Comment (nedrichards) [inline com.github.rhyme_player.rhyme.yml:13]: does wayland actually work here? (try running on the commandline with --nosocket=x11 on a wayland session). If so you should replace x11 with fallback-x11 to use wayland by default.

Comment (gasinvein) [reply com.github.rhyme_player.rhyme.yml:13]: It's Electron. It supports native Wayland, but it's not enabled by default yet.

**Comment by gasinvein** [comment]
Electron complains about OpenGL:
```
[42:0914/115902.019604:ERROR:angle_platform_impl.cc(44)] Display.cpp:878 (initialize): ANGLE Display::initialize error 12289: Could not create a backing OpenGL context.
[42:0914/115902.019705:ERROR:gl_surface_egl.cc(780)] EGL Driver message (Critical) eglInitialize: Could not create a backing OpenGL context.
[42:0914/115902.019740:ERROR:gl_surface_egl.cc(1375)] eglInitialize OpenGL failed with error EGL_NOT_INITIALIZED, trying next display type
[42:0914/115902.020001:ERROR:angle_platform_impl.cc(44)] Display.cpp:878 (initialize): ANGLE Display::initialize error 12289: Could not create a backing OpenGL context.
[42:0914/115902.020036:ERROR:gl_surface_egl.cc(780)] EGL Driver message (Critical) eglInitialize: Could not create a backing OpenGL context.
[42:0914/115902.020064:ERROR:gl_surface_egl.cc(1375)] eglInitialize OpenGLES failed with error EGL_NOT_INITIALIZED
[42:0914/115902.020090:ERROR:gl_initializer_linux_x11.cc(178)] GLSurfaceEGL::InitializeOneOff failed.
[42:0914/115902.020764:ERROR:viz_main_impl.cc(162)] Exiting GPU process due to errors during initialization
[112:0914/115902.030761:ERROR:gpu_init.cc(441)] Passthrough is not supported, GL is swiftshader
```
You probably should add `--device=dri` to `finish-args`.

File: com.github.rhyme_player.rhyme.yml
Comment (Abdallah-Moh) [reply com.github.rhyme_player.rhyme.yml:3]: It was 21.08 but some one asked me to change it

**Comment by Abdallah-Moh** [comment]
bot, build com.github.rhyme_player.rhyme

File: com.github.rhyme_player.rhyme.yml
Comment (gasinvein) [inline com.github.rhyme_player.rhyme.yml:6]: Why did you change it back? `runtime-version` and `base-version` should both be `21.08`.

**Comment by Abdallah-Moh** [comment]
bot, build com.github.rhyme_player.rhyme

File: com.github.rhyme_player.rhyme.yml
Comment (hfiguiere) [reply com.github.rhyme_player.rhyme.yml]: there was no BaseApp for 21.08 at the time, but the problem got addressed and there is a 21.08 baseapp now. So it's good to go.

The problem was the mix and match you did.

File: com.github.rhyme_player.rhyme.yml
Comment (barthalion) [inline com.github.rhyme_player.rhyme.yml:21]:
```suggestion
```

File: com.github.rhyme_player.rhyme.yml
Comment (barthalion) [reply com.github.rhyme_player.rhyme.yml:13]: "not enabled by default" is "not supported" in my book. This should be removed.

File: com.github.rhyme_player.rhyme.yml
Comment (barthalion) [inline com.github.rhyme_player.rhyme.yml:13]:
```suggestion
```

File: com.github.rhyme_player.rhyme.yml
Comment (nedrichards) [reply com.github.rhyme_player.rhyme.yml:13]: Yes, unless you are actually changing electron to support it yourself the permission should be removed for now.

**Comment by nedrichards** [comment]
Other than @barthalion's notes this lgtm

File: com.github.rhyme_player.rhyme.yml
Comment (Abdallah-Moh) [reply com.github.rhyme_player.rhyme.yml:21]: I really cannot see a difference

File: com.github.rhyme_player.rhyme.yml
Comment (Abdallah-Moh) [reply com.github.rhyme_player.rhyme.yml:13]: This is added to support wayland

File: com.github.rhyme_player.rhyme.yml
Comment (hfiguiere) [reply com.github.rhyme_player.rhyme.yml:21]: the suggestion is to remove them, they are not needed.

File: com.github.rhyme_player.rhyme.yml
Comment (hfiguiere) [reply com.github.rhyme_player.rhyme.yml:13]: as said in the review, if it no using wayland by default (ie if the user need to do something) then it doesn't support wayland, therefor the permission should be removed.

**Comment by nedrichards** [comment]
bot, build com.github.rhyme_player.rhyme

**Comment by nedrichards** [comment]
OK, actually after running this again I see that you need to make sure you're declaring all the dbus names the app wants to call so that it integrates properly with the system. Run this with --log-session-bus and identify all the bits that get denied. e.g. this is a symptom of a thing you should probably be allowing:
> [3:0915/213712.509304:ERROR:bus.cc(555)] Failed to get the ownership of org.mpris.MediaPlayer2.chromium.instance3: org.freedesktop.DBus.Error.ServiceUnknown

**Comment by Abdallah-Moh** [comment]
So should I add `--log-session-bus` to finish args

**Comment by gasinvein** [comment]
> So should I add `--log-session-bus` to finish args

No, you should run the app with the option, like `flatpak run --log-session-bus com.github.rhyme_player.rhyme`.

**Comment by Abdallah-Moh** [comment]
bot, build com.github.rhyme_player.rhyme

**Comment by Abdallah-Moh** [comment]
bot, build com.github.rhyme_player.rhyme

**Comment by Abdallah-Moh** [comment]
@nedrichards 
> OK, actually after running this again I see that you need to make sure you're declaring all the dbus names the app wants to call so that it integrates properly with the system. Run this with --log-session-bus and identify all the bits that get denied. e.g. this is a symptom of a thing you should probably be allowing:
> 
> > [3:0915/213712.509304:ERROR:bus.cc(555)] Failed to get the ownership of org.mpris.MediaPlayer2.chromium.instance3: org.freedesktop.DBus.Error.ServiceUnknown

How ?? I tried a search but failed

**Comment by gasinvein** [comment]
I'm not sure if there is a way to select a custom MPRIS player name in electron instead of the automatic one, but if there is none, adding `--own-name=org.mpris.MediaPlayer2.chromium.*` to `finish-args` should make MPRIS work.

**Comment by Abdallah-Moh** [comment]
bot, build com.github.rhyme_player.rhyme

**Pull request closed without merging**
